src/prefect/engine/state.py

```diff
@@ -82,6 +83,37 @@ def __eq__(self, other: object) -> bool:
     def __hash__(self) -> int:
         return id(self)
 
+    def __getstate__(self) -> dict:
+        data = self.__dict__
+
+        # For 'Exception' results, we will check if it can be pickled successfully and
+        # on failure convert the exception to a string instead. This allows the engine
+        # to handle tasks with unpickable exceptions successfully. We must explicitly
+        # ignore state signals here or we will get stuck in a recursive loop because
+        # signals are exceptions that contain a state with `_result` set to the signal
+        if isinstance(self.result, Exception) and not isinstance(
+            self.result, prefect.engine.signals.PrefectStateSignal
+        ):
+
+            # Check for an exception during pickling of the 'Exception' type result
+            pickle_exc: Optional[Exception] = None
```

Comment:
The need for this variable + all the logic below can be brought inside the `Exception` catch

Reply:
Ah yeah now that I've refactored it to handle all exceptions that make sense. Will do.